---
title: 'Human Genome Analysis Lab 3 : Graphics with ggplot2'
author: "Joseph Drake"
date: "2/15/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
SNPs<- read.table("23andMe_complete.txt", header = TRUE, sep = "\t")

```

## Exercise 1

Using ggplot make a make a bar graph of the total SNP counts for each chromosome.


```{r}
ggplot(SNPs, aes(x=chromosome)) + geom_bar()

```

##Exercise 2

Order the chromosomes according to number by converting chromosomes from a factor to a order factor as in the example above. Then replot the bar graph

```{r}

SNPs$chromosome = ordered(SNPs$chromosome, levels=c(seq(1, 22), "X", "Y", "MT"))
ggplot(SNPs, aes(x=chromosome)) + geom_bar()

```


##Exercise 3

Show the contribution of each genotype to the chromosome count using a stacked bar graph (with the fill = genotype)

```{r}

ggplot(SNPs, aes(x=chromosome) ) + geom_bar(aes(fill=genotype))

```

## Exercise 4

Make each set of stacked bars the same height to easier to compare proportions across groups.

```{r}

ggplot(SNPs, aes(x=chromosome)) + 
  geom_bar(aes(fill=genotype), position="fill")
  
```

##Exercise 5

Now place genotypes directly beside one another for each chromosome to compare individual values.

```{r}
ggplot(SNPs, aes(x=chromosome)) + 
  geom_bar(aes(fill=genotype), position="dodge") 
     
```

##Exercise 6

The above graph is pretty hard to read. Try using facet_wrap with the genotype

```{r}

ggplot(SNPs, aes(x=chromosome)) + 
  geom_bar() +
  facet_wrap(.~genotype)

```


